File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for(int num: nums) sum += num;
        if(sum%2==1) return false;
        vector<vector<int>>dp(nums.size()+1,vector<int>(sum/2+1,0));
        for(int index = 1; index <= nums.size(); index++){
            for(int subsetSum = 0; subsetSum <= sum/2; subsetSum++){
                if(subsetSum == 0){
                    dp[index][subsetSum] = 1;
                }else{
                    int skip = dp[index-1][subsetSum];
                    int take = 0;
                    if(subsetSum >= nums[index-1]) take = dp[index-1][subsetSum-nums[index-1]];
                    dp[index][subsetSum] = skip | take;
                }
            }
        }
        
        return dp[nums.size()][sum/2];
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    private boolean subsetSumEqualToTarget(int n,int k,int[] nums){
        boolean[][] dp = new boolean[n][k+1];
        
        for(int i=0;i<n;i++) dp[i][0]=true;
        if(nums[0] <= k) dp[0][nums[0]]=true;
        
        for(int ind=1;ind<n;ind++){
            for(int target=1;target<=k;target++){
                boolean nottake=dp[ind-1][target];
                boolean take=false;
                if(target>=nums[ind]) take=dp[ind-1][target-nums[ind]];
                dp[ind][target]=take|nottake;
            }
        }
        
        return dp[n-1][k];
    }
    public boolean canPartition(int[] nums) {
        int n=nums.length;
        int totalsum=0;
        for(int i=0;i<n;i++){
            totalsum+=nums[i];
        }
        if(totalsum%2 == 1) return false;
        
        int k=totalsum/2;
        
        return subsetSumEqualToTarget(n,k,nums);
    }
}